web_poet/rules.py

```diff
@@ -115,9 +116,66 @@ class ExampleComProductPage(WebPage[Product]):
     """
 
     def __init__(self, *, rules: Optional[Iterable[ApplyRule]] = None):
-        self._rules: List[ApplyRule] = []
+        self._rules: Dict[int, ApplyRule] = {}
+        self._overrides_matchers: Dict[Type[ItemPage], URLMatcher] = defaultdict(
+            URLMatcher
+        )
+        self._item_matchers: Dict[Optional[Type], URLMatcher] = defaultdict(URLMatcher)
+
+        # Ensures that URLMatcher is deterministic in returning a rule when
+        # matching. Currently, `URLMatcher._sort_domain` has this sorting
+        # criteria:
+        #   * Priority (descending)
+        #   * Sorted list of includes for this domain (descending)
+        #   * Rule identifier (descending)
+        # This means that if the priority and domain are the same, the last tie
+        # breaker woulud be the "Rule identifier", this means we can base it on
+        # when a given rule was added to the registry, i.e. a counter.
+        self._rule_counter = 0
+
         if rules is not None:
-            self._rules = list(rules)
+            for rule in rules:
+                self.add_rule(rule)
+
+    def add_rule(self, rule: ApplyRule) -> None:
```

Comment:
There's a potential for returning the `rule_id: int` but this also means that a new `def get_rule(self, rule_id: int)` needs to be defined as well.

Although I don't see an immediate use case to expose this for now apart from potentially storing the `rule_id` somewhere to have a quick registry look-up later on.

Let me know if we should do expose this as early as now.